add axis/axes support to multilevel discrete wavelet transforms

grlee77 opened the pull request:
This PR adds axis support to the multilevel discrete wavelet transforms.  This was enabled by the axis support that was added to _multidim.py in #130 and _dwt.py in #116 .

An example use case is a 3D array where it is desired to apply a multilevel DWT along only a subset of the axes.

I do not allow any non-unique axes to be specified for the 2D and nD transforms.  Also, for simplicity, the same number of transform levels are applied to each axis specified.

grlee77 commented:
I added this because I sometimes only want to transform a subset of the axes of an n-dimensional array, but it also makes implementation of a fully separable variant of `wavedecn`/`waverecn` easy (see #204)

grlee77 commented:
This will need to be rebased on top of #204 once it is merged.

grlee77 commented:
fixed an erroneous statement in the docstring for `wavedec2` and added some other clarifications.  Rebased on latest master

rgommers commented:
this needs a rebase

grlee77 commented:
okay.  ready for review again

rgommers commented:
needs another rebase....

grlee77 commented:
okay.  rebased again

rgommers commented on pywt/tests/test_multilevel.py:
```diff
@@ -447,5 +447,91 @@ def test_array_to_coeffs_invalid_inputs():
 
 
+def test_waverec_axes_subsets():
+    data = np.random.standard_normal((8, 8, 8))
```
you still want to set a random seed here (and in tests below). the actual numbers don't matter, but a rare random failure is always possible with such tests.

rgommers commented on pywt/_multilevel.py:
```diff
@@ -83,13 +86,17 @@ def wavedec(data, wavelet, mode='symmetric', level=None):
 
-    level = _check_level(min(data.shape), wavelet.dec_len, level)
+    try:
+        axes_shape = data.shape[axis]
```
minor thing: with something called "shape" I tend to think it should be a tuple. In ``_check_level`` this is called ``size``, so maybe ``axis_size`` for clarity?

rgommers commented on pywt/_multilevel.py:
```diff
@@ -154,11 +164,14 @@ def wavedec2(data, wavelet, mode='symmetric', level=None):
     [cAn, (cHn, cVn, cDn), ... (cH1, cV1, cD1)] : list
-        Coefficients list
+        Coefficients list.  For user-specified ``axes``, ``cH*``
+        correspond to ``axes[0]`` while ``cV*`` correspond to ``axes[1]``.
```
may also be useful to specify the mapping from list elements to levels here

rgommers commented on pywt/_multilevel.py:
```diff
@@ -273,6 +299,8 @@ def wavedecn(data, wavelet, mode='symmetric', level=None):
         Dxecomposition level (must be >= 0). If level is None (default) then it
         will be calculated using the ``dwt_max_level`` function.
+    axes : sequence of ints, optional
+        Axes over which to compute the DWT. Axes may not be repeated.
```
"Default is None, which means transform all axes"

rgommers commented:
Overall looks good to me, only had a bunch of minor comments.

grlee77 commented:
In addition to addressing the comments above, the most recent commit has a much more detailed description of the coefficient dictionaries returned by `wavedecn`.

grlee77 commented:
Thanks @rgommers for reviewing all of these recent PRs!

rgommers commented:
LGTM now, merging.

> Thanks @rgommers for reviewing all of these recent PRs!

Thank you for sending them:)
